modules/functions/src/controllers/utils.ts

```diff
@@ -0,0 +1,127 @@
+import ConduitGrpcSdk, {
+  ConduitRouteActions,
+  GrpcError,
+  ParsedRouterRequest,
+  RouteBuilder,
+} from '@conduitplatform/grpc-sdk';
+import { FunctionDefinitions, Functions } from '../models';
+import { isNil } from 'lodash';
+import { NodeVM } from 'vm2';
+import { status } from '@grpc/grpc-js';
+
+function getOperation(op: string) {
+  switch (op) {
+    case 'GET':
+      return ConduitRouteActions.GET;
+    case 'POST':
+      return ConduitRouteActions.POST;
+    case 'UPDATE':
+      return ConduitRouteActions.UPDATE;
+    case 'DELETE':
+      return ConduitRouteActions.DELETE;
+    case 'PATCH':
+      return ConduitRouteActions.PATCH;
+    default:
+      return ConduitRouteActions.GET;
+  }
+}
+
+async function executeFunction(
+  call: ParsedRouterRequest,
+  callback: any,
+  functionCode: string,
+  timeout: number,
+  name: string,
+  grpcSdk: ConduitGrpcSdk,
+) {
+  const logs: string[] | undefined = [];
+  const vm = new NodeVM({
+    console: 'redirect',
+    sandbox: {},
+    timeout: timeout,
+    require: {
+      external: true,
+      import: ['lodash'],
+    },
+  });
+  vm.on('console.log', data => {
+    logs?.push(data);
+  });
+
+  let duration;
+  let start;
+  try {
+    const script = `module.exports = function(grpcSdk,req,res) { ${functionCode} }`;
+    start = process.hrtime();
+    const functionInSandbox = vm.run(script);
+    const end = process.hrtime(start);
+    duration = end[0] * 1e3 + end[1] / 1e6;
+    const functionData = functionInSandbox(grpcSdk, call.request, callback);
```

Comment:
the function  ends after line 59 has been executed, but you are measuring duration before.